### PR TITLE
Add advanced version of coordinates widget

### PR DESCRIPTION
**Pull request by ceski-1**
Added by request. I'll fix any conflicts from https://github.com/fabiangreffrath/woof/pull/1759.

### CODE REVIEW

File: src/hu_coordinates.c [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    return sqrt(dx * dx + dy * dy);
+}
+
+static split_fixed_t SplitFixed(fixed_t x)
+{
+    split_fixed_t result;
+
+    result.negative = x < 0;
+    result.base = x >> FRACBITS;
+    result.frac = x & 0xFFFF;
```
Comment (fabiangreffrath) [inline src/hu_coordinates.c:74]: Isn't this `FRACUNIT - 1`?

**Comment by fabiangreffrath** [comment]
Somehow can't get it to work. If I select "Show player coords: advanced", nothing shows up.

File: src/hu_coordinates.c (hunk 1 above)
Comment (ceski-1) [reply src/hu_coordinates.c]: Changed here: https://github.com/fabiangreffrath/woof/pull/1760/commits/1215df0d6754ac47256313cfe376b79890976998

Comment (ceski-1) [reply src/hu_coordinates.c]: > Can we do something like `FRACMASK = FRACUNIT - 1`?

Oh, I didn't see this edit. How about we add a macro for that in a different PR?

**Comment by ceski-1** [comment]
> Somehow can't get it to work. If I select "Show player coords: advanced", nothing shows up.

Maybe I'm doing bad string manipulation here: https://github.com/fabiangreffrath/woof/pull/1760/files#diff-a4f029940e50c96b9e3766ac0da54d42f3a3f0fbd915269dc0b4b4c1c2e3f9f3R97-R98

File: src/hu_coordinates.c (hunk 1 above)
Comment (fabiangreffrath) [reply src/hu_coordinates.c]: Sure, that's ok.

**Comment by fabiangreffrath** [comment]
This works, but well...

```patch
--- a/src/hu_coordinates.c
+++ b/src/hu_coordinates.c
@@ -94,9 +94,10 @@ static split_angle_t SplitAngle(angle_t x)
 
 static void BuildString(hu_multiline_t *const w_coord, char *buf, int len)
 {
-    const char *s = buf;
+    char *s = M_StringDuplicate(buf);
     M_snprintf(buf, len, "%-*s", WIDGET_WIDTH, s);
     HUlib_add_string_keep_space(w_coord, buf);
+    free(s);
 }
 
 static void FixedToString(hu_multiline_t *const w_coord, const char *label,
```

**Comment by ceski-1** [comment]
> This works, but well...

Thanks! Fixed here: https://github.com/fabiangreffrath/woof/pull/1760/commits/566e848d4f9e73afbb88436deb2c47c837df0c59

**Comment by fabiangreffrath** [comment]
There must be a better solution than running `strdup()` several times per tic on the same string.

**Comment by ceski-1** [comment]
> There must be a better solution than running `strdup()` several times per tic on the same string.

Different approach here: https://github.com/fabiangreffrath/woof/pull/1760/commits/bc8e19b48687f8592b21778612d2d464df0e0e18

**Review by fabiangreffrath — APPROVED** [review]
Much better, thank you!

**Pull request merged**